docs/blog/migrating-the-ecosystem-to-esm.md

```diff
@@ -0,0 +1,140 @@
+---
+title: Migrating the ecosystem to ES modules
+author:
+  - name: The e18e Contributors
+sidebar: false
+date: 2025-03-19
+head:
+  - - meta
+    - property: og:type
+      content: website
+  - - meta
+    - property: og:title
+      content: Migrating the ecosystem to ES modules
+  - - meta
+    - property: og:url
+      content: https://e18e.dev/blog/migrating-ecosystem-to-es-modules
+  - - meta
+    - property: og:description
+      content: An update on the progress we're making migrating packages to ES modules
+---
+
+_March 19, 2025_
+
+Node made some huge steps forward recently by releasing `require(esm)` to each of the long-term support versions (other than the soon to fall out of LTS 18.x). Awesome work by [@joyeecheung](https://bsky.app/profile/joyeecheung.bsky.social) getting it over the line!
+
+This is awesome work and unblocks a huge amount of packages on the migration path from CommonJS to ES modules :tada:
+
+So let's have a look into what's next and how the e18e community is trying to help!
+
+## `require(esm)`
+
+One of the main blockers in the CJS vs ESM story has been interop. You could import CJS from within an ES module, but could not `require` an ES module from within a CJS module.
+
+This left us with two options at the time for a migration path:
+
+- Change much of the code to be `async`, such that we can use a dynamic `import` (which does work in CJS modules)
+- Change all CJS dependencies of our ES module package to be ES modules
+
+The ideal is the latter, that everyone uses ES modules. However, this obviously isn't feasible since not all packages are maintained and it'd be a crazy amount of work to replace them.
+
+So we were blocked...
+
+That is until `require(esm)` came along! This basically means we _can_ `require` an ES module inside a CJS module now. CJS packages can consume ES module packages and vice versa. A huge move forward and unblocks us all to get back on the migration path!
+
+## Types of package
+
+As part of this migration, we have three types of package to deal with:
+
+- ESM packages
+- CommonJS packages
+- Dual packages
+
+### Migrating a CommonJS package
+
+In most cases, this is as simple as setting `type` to `"module"` in `package.json`, then updating all imports to include file extensions.
+
+For example:
+
+```json
+{
+  "name": "my-package",
+  "type": "module"
+}
+```
+
+### Migrating a dual package
+
+Dual packages are basically a necessary evil if you want to support both CJS and ESM _before `require(esm)` was available_.
```

Comment:
this isn't quite accurate; a dual package isn't necessary to support both CJS and ESM because you can `import` CJS. The reasons for a dual package are far more nuanced, and don't actually exist for most packages that have shipped dual.

Reply:
this is true only if they insist on writing native ESM, though - that's the missing nuance here, i think. iow, shipping CJS-only is a trivial way to avoid any need for a dual package (prior to `require(esm)`)

Reply:
that's also not quite true. various packages ship dual so they can support browsers natively without the need for a bundler. they had to use ES modules because browsers don't understand CJS, so they had to ship dual to also support older Node (or require a dynamic import).

i think this shows there's too many edge cases and situations here to mention without going off on a huge tangent

Reply:
that's true.

maybe just say "support both CJS and non-node ESM"?